src/mutation.cpp

```diff
@@ -12,98 +12,308 @@
+        remove_mutation(flag);
     } else {
-        set.erase( toggled_element );
-        trait_keys.erase(str);
+        // We found something to downgrade to, figure out which of the downgrades
+        // are valid things to downgrade to.
+        std::vector<muttype_id> valid_downgrades;
+        for (auto &i : mut_types[flag].get_replaces()) {
+            // First check if we have an opposite mutation
+            if (has_conflicting_mut(i)) {
+                // Not a valid thing to downgrade to, continue
+                continue;
+            }
+            
+            // Then check if we have all the prereqs for a downgrade
+            bool valid = true;
+            for (auto &check : i.get_prereqs()) {
+                if (!has_mut(check)) {
+                    // Missing a prereq!
+                    valid = false;
+                    break;
+                }
+            }
+            // Check again for prereqs2 if needed
+            if (valid) {
+                for (auto &check : i.get_prereqs2()) {
+                    if (!has_mut(check)) {
+                        // Missing a prereq2!
+                        valid = false;
+                        break;
+                    }
+                }
+            }
+            
+            // If it's still valid, add it to the list
+            if (valid) {
+                valid_downgrades.push_back(i);
+            }
+        }
+        
+        // Hopefully at this point we have a valid downgrade
+        muttype_id replacement;
+        if (!valid_downgrades.epmty()) {
+            // Pick a random valid downgrade
+            replacement = valid_downgrades[rng(0, valid_downgrades.size() - 1)]
+        } else {
+            // If we don't have any valid downgrades, force a random downgrade
+            std::vector<muttype_id> choices = mut_types[flag].get_replaces();
+            replacement = choices[rng(0, choices.size() - 1);
+        }
+        // Replace our old mutation with the new one.
+        replace_mutation(flag, replacement);
     }
 }
 
-int Character::get_mod(std::string mut, std::string arg) const
+int mutation_type::get_mod(std::string arg, bool active) const
 {
-    auto &mod_data = mutation_data[mut].mods;
     int ret = 0;
-    auto found = mod_data.find(std::make_pair(false, arg));
-    if (found != mod_data.end()) {
+    auto found = mods.find(std::make_pair(active, arg));
+    if (found != mods.end()) {
         ret += found->second;
     }
-    /* Deactivated due to inability to store active mutation state
-    if (has_active_mutation(mut)) {
-        found = mod_data.find(std::make_pair(true, arg));
-        if (found != mod_data.end()) {
-            ret += found->second;
-        }
-    } */
     return ret;
 }
 
-void Character::apply_mods(const std::string &mut, bool add_remove)
+void Character::apply_mods(const muttype_id &mut_id, bool add_remove, bool active)
 {
+    auto mut = &mut_types[mut];
```

Comment:
This should be `const auto &mut = mut_types[mut_id]`, you want a const reference to the entry in the map. The current version gets a pointer to the entry in the map.